### PR TITLE
eight schools example

### PR DESCRIPTION
**Pull request by drohde**
This is an implementation and a write up of the eight schools example - including a comparison between NUTS/Stan Edward/KLqp and Edward/HMC.  I think it is on the way to being a useful example and tutorial (ipython notebook is also included).

I first mentioned this here: https://github.com/blei-lab/edward/issues/805

It is fairly complete, but it will need some polishing before merging - I am at the point where I would like some feedback.

I have had trouble tuning the ed.HMC call and I currently just use the default parameters with probably too many samples.

@dustinvtran

### CODE REVIEW

File: examples/eight_schools/eight_schools.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
+"""implement the stan 8 schools example
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: Can you rephrase as: `Implement the 8 schools example using the recommended non-centered parameterization.:.`

File: examples/eight_schools/eight_schools.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+using the recommended non-centred parameterisation
+
+https://github.com/stan-dev/rstan/wiki/RStan-Getting-Started
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: Can you move links to bottom of docstring?

Followed by References subsection, e.g.,
```
#### References

http://...
```

File: examples/eight_schools/eight_schools.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+http://mc-stan.org/users/documentation/case-studies/divergences_and_bias.html
+we slightly modify the stan example to:
+                                         avoid improper priors
+                                         avoid half-cauchy priors
+make the statement of the model directly comparable
+
+This model has a hierachy and an inferred variance - yet the example is
+very simple - only the Normal distribution is used"""
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: Last docstring `"""` should be at newline. All sentences should read like English (e.g., periods; grammatical sentences).

File: examples/eight_schools/eight_schools.stan [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+model {
+  mu ~ normal(0, 10); 
+  logtau ~ normal(5, 1);
+  theta_tilde ~ normal(0, 1);
+  for (j in 1:J) {
+      y[j] ~ normal(mu + exp(logtau) * theta_tilde[j], sigma[j]);
+  }
+}
+
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.stan]: delete blankline

File: examples/eight_schools/eight_schools.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+    # ed.KLqp inference
+    q_logtau = Normal(tf.Variable(tf.random_normal([1])),
+                          tf.nn.softplus(tf.Variable(tf.random_normal([1]))))
+    q_mu = Normal(tf.Variable(tf.random_normal([1])),
+                      tf.nn.softplus(tf.Variable(tf.random_normal([1]))))
+    q_theta_tilde = Normal(tf.Variable(tf.random_normal([J])),
+                               tf.nn.softplus(tf.Variable(tf.random_normal([J]))))
+    par2q = {logtau: q_logtau, mu: q_mu, theta_tilde: q_theta_tilde}
+    inference = ed.KLqp(par2q, data=thedata)
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: As part of best practices, can you write the args inline if they're not reused? That is,
```python
inference = ed.KLqp({logtau: q_logtau, ...}, 
                    data=thedata)
```

File: examples/eight_schools/eight_schools.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import tensorflow as tf
+from edward.models import Normal, Empirical
+import numpy as np
+
+def main(_):
+    # data
+    J = 8
+    data_y = np.array([28, 8, -3, 7, -1, 1, 18, 12])
+    data_sigma = np.array([15, 10, 16, 11, 9, 11, 10, 18])
+    # end data
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: rm end data comment

File: examples/eight_schools/eight_schools.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+    # model definition
+    mu = Normal(tf.zeros([1]), tf.fill([1], 10.))
+    logtau = Normal(tf.fill([1], 5.), tf.fill([1], 1.))
+    theta_tilde = Normal(tf.zeros([J]), tf.ones([J]))
+    sigma = tf.placeholder(tf.float32, J)
+    y = Normal(mu + tf.exp(logtau) * theta_tilde, sigma * tf.ones([J]))
+    # end model definition
+    
+    thedata = {y: data_y, sigma: data_sigma}
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: Instead of `thedata`, rename to just `data`?

File: examples/eight_schools/eight_schools.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+def main(_):
+    # data
+    J = 8
+    data_y = np.array([28, 8, -3, 7, -1, 1, 18, 12])
+    data_sigma = np.array([15, 10, 16, 11, 9, 11, 10, 18])
+    # end data
+
+    # model definition
+    mu = Normal(tf.zeros([1]), tf.fill([1], 10.))
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: If `mu` is scalar, you can also just use
```python
mu = Normal(0., 10.)
```
Same goes for `logtau`.

File: examples/eight_schools/eight_schools.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    logtau = Normal(tf.fill([1], 5.), tf.fill([1], 1.))
+    theta_tilde = Normal(tf.zeros([J]), tf.ones([J]))
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: For vector shapes, can just use `tf.zeros(J)` and `tf.ones(J)`.

File: examples/eight_schools/eight_schools.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    sigma = tf.placeholder(tf.float32, J)
+    y = Normal(mu + tf.exp(logtau) * theta_tilde, sigma * tf.ones([J]))
+    # end model definition
+    
+    thedata = {y: data_y, sigma: data_sigma}
+
+    # ed.KLqp inference
+    q_logtau = Normal(tf.Variable(tf.random_normal([1])),
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: Instead of `tf.Variable`, best practice is to use
```python
tf.get_variable("q_logtau", [])
```

File: examples/eight_schools/eight_schools.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    S = 400000
+    burn = S // 2
+    hq_logtau = Empirical(tf.Variable(tf.zeros([S, 1])))
+    hq_mu = Empirical(tf.Variable(tf.zeros([S, 1])))
+    hq_theta_tilde = Empirical(tf.Variable(tf.zeros([S, J])))
+
+    par2hq = {logtau: hq_logtau, mu: hq_mu, theta_tilde: hq_theta_tilde}
+    inference = ed.HMC(par2hq, data=thedata)
+    inference.run()
+    # end HMC inference
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: remove `# end ...` comments

File: examples/eight_schools/eight_schools.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    print("E[mu] = %f" % (hq_mu.params.eval()[burn:].mean()))
+    print("E[logtau] = %f" % (hq_logtau.params.eval()[burn:].mean()))
+    print("E[theta_tilde]=")
+    print(hq_theta_tilde.params.eval()[burn:, ].mean(0))
+    print("====  end ed.HMC inference ====")
+    print("")
+    print("")
+
+    # Stan if Pystan is installed
+    try:
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: As an individual Edward example, maybe move this to its own script? Also, we probably shouldn't store R scripts. So we can replace the Rstan script with the Pystan script.

File: examples/eight_schools/eight_schools.py (hunk 8 above)
Comment (drohde) [reply examples/eight_schools/eight_schools.py]: I seem to need to use: 
mu = Normal([0.], [1.])

**Comment by drohde** [comment]
Thanks @dustinvtran, I have made the updates.  I noticed there were a lot of pep8 conflicts.  I think I fixed all the ones in the python files, I found it difficult to resolve some issues in the notebook.

File: examples/eight_schools/eight_schools.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+
+def main(_):
+    # data
+    J = 8
+    data_y = np.array([28, 8, -3, 7, -1, 1, 18, 12])
+    data_sigma = np.array([15, 10, 16, 11, 9, 11, 10, 18])
+
+    # model definition
+    mu = Normal([0.], [10.])
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: Why not `mu = Normal(0., 10.)`?

File: examples/eight_schools/eight_schools.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+    print("E[theta_tilde]=")
+    print((q_theta_tilde.mean().eval()))
+    print("====  end ed.KLqp inference ====")
+    print("")
+    print("")
+
+    # HMC inference
+    S = 400000
+    burn = S // 2
+    hq_logtau = Empirical(tf.Variable(tf.zeros([S, 1])))
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: `tf.get_variable`

File: examples/eight_schools/eight_schools_pystan.py [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+import pystan
+
+
+def main():
+    # data
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools_pystan.py]: 2-space indent

File: examples/eight_schools/eight_schools.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+from __future__ import print_function
+
+import edward as ed
+import tensorflow as tf
+import numpy as np
+from edward.models import Normal, Empirical
+
+
+def main(_):
+    # data
```
Comment (dustinvtran) [inline examples/eight_schools/eight_schools.py]: We use 2-space indent (https://www.tensorflow.org/community/style_guide).

File: notebooks/eight_schools.ipynb [hunk 17]
```diff
@@ -0,0 +1,533 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Eight Schools - A Stan/Edward Comparison\n",
+    "\n",
+    "In this tutorial we fit the 8 schools model to data made famous by __[Stan](https://github.com/stan-dev/rstan/wiki/RStan-Getting-Started)__,  __[Bayesian Data Analysis](http://www.stat.columbia.edu/~gelman/book/)__ and first appearing in __[Rubin (1981) - (paywalled)](http://journals.sagepub.com/doi/abs/10.3102/10769986006004377)__.  The tutorial will be particularly suited to Stan users who are familiar with hierachichal models that would like to start using Edward.\n",
```
Comment (dustinvtran) [inline notebooks/eight_schools.ipynb]: s/hierachichal/hierarchical

File: notebooks/eight_schools.ipynb [hunk 18]
```diff
@@ -0,0 +1,533 @@
...
+    "\n",
+    "\n",
+    "In this example we introduce a slightly modified 8 Schools example and we compare Stan using the No-U-Turn sampler (NUTS) and Edward using Automatic Differential Variational Bayes (ADVI/KLQP) and Hamiltonian Monte Carlo (HMC).  Both the Stan and Edward code are included in this tutorial and you may choose to install __[pystan](http://pystan.readthedocs.io/en/latest/getting_started.html)__ or __[rstan](https://github.com/stan-dev/rstan/wiki/RStan-Getting-Started)__ to run the Stan portions of code.  Comparison of Stan and Edward implementations can be a good way to check the algorithm and model is correct.  As we use our own Stan implementation of 8 schools our results will differ slightly with the improper priors used in the Stan __[getting started] (https://github.com/stan-dev/rstan/wiki/RStan-Getting-Started)__ page and the Normal/Cauchy priors used in __[Diagnosing Biased Inference with Divergences](http://mc-stan.org/users/documentation/case-studies/divergences_and_bias.html)__.  \n",
```
Comment (dustinvtran) [inline notebooks/eight_schools.ipynb]: s/Automatic Differential Variational Bayes/Automatic Differentiation Variational Inference

File: notebooks/eight_schools.ipynb [hunk 19]
```diff
@@ -0,0 +1,533 @@
...
+    "\n",
+    "We use the non-centred parameterisation as advocated in __[Diagnosing Biased Inference with Divergences](http://mc-stan.org/users/documentation/case-studies/divergences_and_bias.html)__ - which is highly recommended reading, just note all three models use slightly different priors when comparing results.\n",
+    "\n",
+    "In the centred (and simplified) form eight schools is a hierarchichal model:\n",
```
Comment (dustinvtran) [inline notebooks/eight_schools.ipynb]: s/centred/centered

(this typo appear below too)

File: notebooks/eight_schools.ipynb [hunk 20]
```diff
@@ -0,0 +1,533 @@
...
+    "\n",
+    "$y_j \\sim \\mathcal{N}(\\theta, \\sigma^2_j)$\n",
+    "\n",
+    "where $j \\in [1..8]$ and $\\{ y_j, \\sigma_j \\}$ are observed.\n",
+    "\n",
+    "It may be easier to think of $\\tau$ as the root variance, which we give a log normal prior to.\n",
+    "\n",
+    "\n",
+    "\n",
+    "The Non-centred model is equivelent and more suitable for inference and takes the following form:\n",
```
Comment (dustinvtran) [inline notebooks/eight_schools.ipynb]: s/equivelent/equivalent

File: notebooks/eight_schools.ipynb [hunk 21]
```diff
@@ -0,0 +1,533 @@
...
+    "fit = pystan.stan(model_code=eight_schools_stan, \n",
+    "                  data=standata, iter=100000)\n",
+    "print(fit)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The inferred parameter theta_tilda or $\\theta'$ is a transform, we might prefer to convert this back to the more intepretable theta by posthoc processing of the samples or by including a generated quantities block in the Stan code.  In this case the required postprocessing is quite simple."
```
Comment (dustinvtran) [inline notebooks/eight_schools.ipynb]: Does $\\tilde{\\theta}$ work? Currently you call it theta_tilde but use the symbol for "theta prime".

File: notebooks/eight_schools.ipynb [hunk 22]
```diff
@@ -0,0 +1,533 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Eight Schools - A Stan/Edward Comparison\n",
```
Comment (dustinvtran) [inline notebooks/eight_schools.ipynb]: Since this notebook is in Edward: maybe rename to "Edward/Stan", and move the Stan section below the Edward section?

File: examples/eight_schools/eight_schools.py (hunk 13 above)
Comment (drohde) [reply examples/eight_schools/eight_schools.py]: I get a different result with the two calls:

In [4]: Normal(0., 10.)
Out[4]: <ed.RandomVariable 'Normal/' shape=() dtype=float32>

In [5]: Normal([0.], [10.])
Out[5]: <ed.RandomVariable 'Normal_1/' shape=(1,) dtype=float32>

.. am I missing something?

File: notebooks/eight_schools.ipynb (hunk 21 above)
Comment (drohde) [reply notebooks/eight_schools.ipynb]: It isn't great, but it is better.. thanks!

File: examples/eight_schools/eight_schools.py (hunk 13 above)
Comment (dustinvtran) [reply examples/eight_schools/eight_schools.py]: Nope. It being a scalar random variable is what I'm suggesting. :)

This more closely aligns with the Stan program, which uses scalars for the group-level parameters `mu` and `logtau`. A vector of length one adds slightly additional complexity.

**Comment by drohde** [comment]
Thanks for your comments @dustinvtran the notebook could have more changes to reflect the reorganisation, but I ran out of time today and just pushed what I got done.

As I mentioned, changing the mu and a logtau to a scalar isn't working for me, I will look into why next week.

**Comment by dustinvtran** [comment]
Got it. Thanks for spending your time on this. This is certainly a very useful contribution.

**Comment by drohde** [comment]
@dustinvtran I have got on top of the pep8 violations, but the build still fails due to the import pystan.

**Comment by dustinvtran** [comment]
Can you include `pystan` in Travis' pip installs? https://github.com/blei-lab/edward/blob/master/.travis.yml#L37

**Comment by drohde** [comment]
@dustinvtran ok, after reducing the number of iterations it seems to have passed

**Comment by dustinvtran** [comment]
Thanks again!

**Comment by drohde** [comment]
@dustinvtran thank you!

Do you need a tex file to make it complete?

**Comment by dustinvtran** [comment]
Yep. We haven't automated that step yet.

**Comment by drohde** [comment]
@dustinvtran I couldn't work out how to compile the latex - it seemed to be a fragment without a header...

 I assume you use the tex to generate the html at some point..

**Comment by dustinvtran** [comment]
Yep. Check out the set up process in https://github.com/blei-lab/edward/tree/master/docs
